File: backend/core/network/protocols.py
"""SoAI - Core network protocol contracts [backend/core/network/protocols.py]"""
# SPDX-License-Identifier: LicenseRef-SoAI-Source-1.0

from __future__ import annotations

from typing import Protocol

from core.types.json import JSONValue

__all__ = ("HTTPJsonResponseProtocol",)


class HTTPJsonResponseProtocol(Protocol):
    def json(self) -> JSONValue: ...

File: frontend/assets/ts/core/search/protocols.ts
/* SoAI - Shared search protocols [frontend/assets/ts/core/search/protocols.ts] */
// SPDX-License-Identifier: LicenseRef-SoAI-Source-1.0

import type { SearchIndex, SearchItem } from '@core/search/searchTypes.ts';
import type { JsonValue } from '@core/types/jsonValues.ts';

const SEARCH_PANEL_SERVICE_ID = 'features.search.panel';
const SEARCH_SOURCE_HARDWARE = 'hardware';

type SearchIndexContributorBucket = keyof SearchIndex;

interface SearchIndexContributor {
    readonly id: string;
    readonly bucket: SearchIndexContributorBucket;
    readonly source: string;
    index: (resourceValue: JsonValue) => SearchItem[];
}

interface SearchStaticIndexContext {
    readonly grantedActions: ReadonlySet<string>;
}

interface SearchStaticIndexContributor {
    readonly id: string;
    readonly bucket: SearchIndexContributorBucket;
    index: (context: SearchStaticIndexContext) => SearchItem[] | Promise<SearchItem[]>;
}

export { SEARCH_PANEL_SERVICE_ID, SEARCH_SOURCE_HARDWARE };
export type { SearchIndexContributor, SearchIndexContributorBucket, SearchStaticIndexContext, SearchStaticIndexContributor };
